### PR TITLE
expose the last evaluated result as `it` in REPL

### PR DESCRIPTION
**Pull request by valyagolev**
This depends on #733.

This partially addresses #304.

There's one essential commit with the functionality: 8666e55d104821a5b026eb4b5da031c563e6462d . And then there's a small refactoring that I'm proposing, but since I'm unfamiliar with the codebase, I have no idea if it's useful or not.

### CODE REVIEW

**Review by xsebek — COMMENTED** [review]
The code and even the refactoring seem reasonable to me.

I just have no idea how the `it` variable works in the code. I guess you grab the result from the finished REPL computation and set it in the base robot context?

**Comment by byorgey** [comment]
@valyagolev , since this has merge conflicts, and now that you have push access, it's probably easiest to just close this PR and create a new one from a branch in the `swarm` repo itself rather than from a branch in your fork.  (I'll assume you know what that means but feel free to ask for clarification/help if not.)

**Comment by xsebek** [comment]
@byorgey in this case the conflict is small enough and should be easy to resolve.

IIRC the only reason for creating branches in the swarm repo was the Restyled Action worked better that way or was there another reason?

**Comment by valyagolev** [comment]
It just needs a rebase :) The branch is the continuation of the one
squash-merged. I will do the rebase when I wake up.

Thank you for the welcome!

**Comment by byorgey** [comment]
Yes, I'm sure fixing the merge conflicts will be easy, I just meant if some "administrative" work (such as rebasing) has to happen anyway, it could be an opportunity to reopen a PR from a swarm branch.  Not required, just a suggestion.  

> IIRC the only reason for creating branches in the swarm repo was the Restyled Action worked better that way or was there another reason?

That is one reason, but the other reason is that it makes collaboration easier.  For example #669 was a very collaborative effort, and that is harder when PRs are opened from people's private forks where no one else can push.  (Not that we are going to immediately start pushing commits to others' PRs!  But as trust develops it becomes a possibility.)

**Comment by valyagolev** [comment]
I kept this branch & PR for this one to avoid duplicating the conversation. But next time I'll gladly push a new branch to the main repo directly. Thank you! 

The conflict and the comments were resolved.

File: src/Swarm/Game/Robot.hs
Comment (byorgey) [inline src/Swarm/Game/Robot.hs:436]: Can we put a blank line in between `runningAtomic` and `emptyRobotContext`?  A Haddock comment might be nice too, though I guess `emptyRobotContext` is pretty self-explanatory.

File: src/Swarm/Language/Pipeline.hs
Comment (byorgey) [inline src/Swarm/Language/Pipeline.hs:42]: What types will end up being substituted for `val`?  I think I saw `Value` and `Maybe Value`, will there be others?

I wonder if we can come up with a better name than `Processed`.  To me, the name doesn't really help me guess what the definition is.  I guess you named it in parallel with `ProcessedTerm`, but a `ProcessedTerm` actually results from processing a `Term`, whereas that's not the case for a `Processed Value`. (Though perhaps we should rename `ProcessedTerm` as well!)  Also, seeing the type `Processed` made me expect that `ProcessedTerm` is now a type synonym for `Processed Term` or something like that, but it's not.

`Decorated`, perhaps?  `Typeful`?  `Typed`?  I'm open to suggestions.

File: src/Swarm/Game/Robot.hs
Comment (byorgey) [inline src/Swarm/Game/Robot.hs:516]: I like the idea of adding `Ixed` and `At` instances for `RobotContext`, but I would tend to put them right next to the definition of `RobotContext`.  Was there a particular reason for having them down here?

File: src/Swarm/Game/Robot.hs
Comment (byorgey) [inline src/Swarm/Game/Robot.hs:72]: I think this should be listed in the "Robot context" section earlier in the export list.

File: src/Swarm/Game/Robot.hs
Comment (byorgey) [reply src/Swarm/Game/Robot.hs:436]: Also, is there a particular reason for putting `emptyRobotContext` here?  I would tend to put it near the definition of `RobotContext`.

File: src/Swarm/TUI/Controller.hs
Comment (byorgey) [inline src/Swarm/TUI/Controller.hs:713]: Love it, I knew this code could be simplified somehow.

**Review by byorgey — COMMENTED** [review]
This is looking nice!  I just have a few comments about cosmetic issues.

We should probably also talk about the `it` variable in the tutorial, but we can make a separate issue for that.

File: src/Swarm/Language/Pipeline.hs
Comment (valyagolev) [reply src/Swarm/Language/Pipeline.hs:42]: > What types will end up being substituted for val? I think I saw Value and Maybe Value, will there be others?

I'd also not be too surprised if I saw there: `()` or `Term`; `[Var]` (e.g. as a list of suggestions for a typed hole)... It has a functorial feel to it. Obviously I'm not familiar enough with the codebase to have any idea if it's good or not. I also thought maybe there could be a way to have `ProcessedTerm` be part of this already, but I'm not sure.

> I wonder if we can come up with a better name than Processed. 

Completely agree – it's more of a placeholder name, I was hoping for a feedback.

Actually I'm wondering why is it that there's such a hard separation between `Type` and `Requirements`. I guess they "mean" different things, but are they actually ever separated from each other anyway? Doesn't their calculation follow similar paths? Wouldn't they be essentially a decoration over a type in other PLs?

So with that in mind, I do like `Typed`. 

Maybe there could be a name for a generalised type like this, e.g... `AssemblableType`? 

then we could have:

    data AssemblableType = AssemblableType Requirements Polytype
    data Assemblable v = Assemblable AssemblableType v

or even:

    data Assemblable v = AssemblableType Requirements Polytype v
    type AssemblableType = Assemblable ()

And yeah, whatever we decide about the meaning of this should determine which file it goes to, as well.

File: src/Swarm/Game/Robot.hs
Comment (valyagolev) [reply src/Swarm/Game/Robot.hs:516]: I'm just a bit lost in big files. Also I think the best placement for everything will get clearer if we clarify the specific types proposed here... 

I'm also not very sure if I was right not to include the contents of the `Store`...

There's another way to approach this particular change, namely by changing `RobotContext` into a product of `Ctx (Typed Value)` (or whatever we agree on) and the `Store`. I do think the three `Ctx`s should go together somehow. Three different maps potentially mean three different sets of `Var`s: do we want them?

(Such a change will remove the need for those instances)

File: src/Swarm/Language/Pipeline.hs
Comment (xsebek) [reply src/Swarm/Language/Pipeline.hs:42]: I like `Typed`, that sounds most natural to me. I might consider using a record though:
```Haskell
data Typed v = Typed
  { value :: v,
  , typed :: Polytype
  , requires :: Requirements
  }
```
We use lenses a lot so we might as well use them for this too.


Btw. if you want to get feedback on something in PR, it is a good idea to insert some `TODO` comments so that we don't forget about it. Alternatively, you can make comments on the code here on GitHub or in your editor if you have some plugin for that. :wink:

File: src/Swarm/Game/Robot.hs
Comment (xsebek) [reply src/Swarm/Game/Robot.hs:516]: If you have any suggestions on how to split the Robot source file to be easier to orient yourself, feel free to chime in:
- #707 :wink: 

I did not mention this file in the list because it is only over 500 loc, but it is still in our top 10. :grinning:

File: src/Swarm/Language/Pipeline.hs
Comment (byorgey) [reply src/Swarm/Language/Pipeline.hs:42]: > Actually I'm wondering why is it that there's such a hard separation between Type and Requirements. I guess they "mean" different things, but are they actually ever separated from each other anyway? Doesn't their calculation follow similar paths? Wouldn't they be essentially a decoration over a type in other PLs?

Yes, very good observation.  Right now requirements are calculated separately, after typechecking, but actually there are several bugs related to the fact that it doesn't work quite right (#540, #394) and I want to redesign things so that requirements are part of types, as they should be (#231).

But let's not get too crazy with refactoring in this PR (what you have so far is fine).  I'd rather merge something that works to give us `it`, and then contemplate further refactoring as a separate PR.

File: src/Swarm/Game/Robot.hs
Comment (byorgey) [reply src/Swarm/Game/Robot.hs:516]: > There's another way to approach this particular change, namely by changing RobotContext into a product of Ctx (Typed Value) (or whatever we agree on) and the Store.

Yes, we contemplated this previously, in #190 .  We abandoned it because it got too complex.  But maybe if we try again we can find a simpler way to do it.

**Comment by byorgey** [comment]
@valyagolev This is looking good so far.  Let us know if you feel like you need more guidance to know how to proceed from here.

**Comment by valyagolev** [comment]
what's the verdict on `Processed`? I guess I could just rename it to `Typed` and extract to its own module for now

i will apply other suggested changes soon and merge. thanks!

**Comment by xsebek** [comment]
I think we agreed on `Typed`, so please do so 🙂

**Comment by valyagolev** [comment]
added 9ca4e4666f5adf68125e13063d0e8050020d23b7 to fix #304 
Screenshot 2022-10-11 at 11 05 26

File: src/Swarm/Language/Typed.hs
Comment (byorgey) [inline src/Swarm/Language/Typed.hs:12]:
```suggestion
-- | A value, or a hole, or something else that has its type & requirements fixed
```

File: src/Swarm/Game/State.hs
Comment (byorgey) [inline src/Swarm/Game/State.hs:370]: Eventually I want to stop assuming that the base is always ID 0, and this refactoring is a great first step!

**Review by byorgey — COMMENTED** [review]
This doesn't seem to work quite right when executing commands.  For example, if I type `build {move}` at the prompt, it says `it1  :=  <r1> : robot`, which makes sense.  But then when I type `it1` at the prompt, it says it has type `cmd robot` instead of `robot`.  And if I try to actually use `it1`, it causes a fatal crash.  We do have some special case behavior built in for things of `cmd` type entered at the REPL, and I suspect we're running into that somehow, though I haven't yet figured out what's going on exactly.

We should also reset the `replNextValueIndex` to 0 every time we start a new game (see the `startGameWithSeed` function in `Swarm.TUI.Model`).

**Comment by valyagolev** [comment]
> This doesn't seem to work quite right when executing commands. For example, if I type build {move} at the prompt, it says it1 := <r1> : robot, which makes sense. But then when I type it1 at the prompt, it says it has type cmd robot instead of robot. And if I try to actually use it1, it causes a fatal crash. We do have some special case behavior built in for things of cmd type entered at the REPL, and I suspect we're running into that somehow, though I haven't yet figured out what's going on exactly.

Oh, this I didn't test. Yeah I will test a few weirder scenarios, maybe we'll need a test for those...

File: CONTRIBUTING.md
Comment (xsebek) [inline CONTRIBUTING.md:223]: I was not aware you also added it here, @valyagolev. :laughing: Meanwhile, I opened #744.

Technically this is not true - because of our restyled configuration file from times of Fourmolu 0.3 it runs with:
```bash
find src/ app/ test/ -name "*.hs" | xargs fourmolu --mode=inplace '--ghc-opt=-XBangPatterns' \
         '--ghc-opt=-XFlexibleContexts' \
         '--ghc-opt=-XFlexibleInstances'\
         '--ghc-opt=-XGADTSyntax'\
         '--ghc-opt=-XMultiParamTypeClasses'\
         '--ghc-opt=-XNumericUnderscores'\
         '--ghc-opt=-XRankNTypes'\
         '--ghc-opt=-XTupleSections'\
         '--ghc-opt=-XTypeApplications'\
         '--ghc-opt=-XTypeOperators'\
         '--ghc-opt=-XImportQualifiedPost'\
         '--ghc-opt=-XLambdaCase'
```
Which should be the same, but I should probably update it anyway.

While it is nice that you noted it down, it's kind of silly to put everything in this one PR, please feel welcome to open other small PRs. :wink:

Comment (valyagolev) [reply CONTRIBUTING.md:223]: oh that explains why i had to manually undo some of its changes then :)

I just copied what I was suggested. I will remove it, thanks for your PR

**Comment by valyagolev** [comment]
> [9ca4e46](https://github.com/swarm-game/swarm/commit/9ca4e4666f5adf68125e13063d0e8050020d23b7) to fix #304 Screenshot 2022-10-11 at 11 05 26

are we okay with this syntax?

**Comment by xsebek** [comment]
I think `it0: int = 1` would look more like the def/let syntax we use currently, but I don't have a strong opinion.

File: CONTRIBUTING.md
Comment (xsebek) [reply CONTRIBUTING.md:223]: Yes, you will have to revert the change to this file, because I merged first, sorry about that. :sweat_smile: 

You can do that automatically but in this case, it's quite easy either way.

**Comment by valyagolev** [comment]
>  We do have some special case behavior built in for things of cmd type entered at the REPL, and I suspect we're running into that somehow, though I haven't yet figured out what's going on exactly.

Ok, yes, now I see the problem and it requires making a decision.

Basically if I run a line that returns `cmd a`, REPL will run the line to get `cmd a`, then run the command to return `a` (e.g. a particular `robot`). So the question is: do we want to store the `cmd a` or `a` in `it*`?

If we store `cmd a`, we will recreate (e.g.) a robot every time we run `it`. Because one of the main use-cases is to avoid losing references to the robots, probably it's better to store `a`. `cmd a` is available through the repl history anyway, I suppose.

ghci stores `a`:
```
λ> print "hello" >> return 3
"hello"
3

λ> it
3
```

**Comment by valyagolev** [comment]
Screenshot 2022-10-15 at 12 18 30

This looks better

**Comment by byorgey** [comment]
> So the question is: do we want to store the `cmd a` or `a` in `it*`?

Yes, absolutely we want to store the returned `a` value, just like `ghci` does.

By the way, relatedly, the syntax
```
it4 : robot <- <r2>
```
looks strange, because `<r2>` itself does not have a `cmd` type, so it's not something you could bind with `<-`.  It should just be displayed with an `=` sign like everything else.  The fact that we executed a command to get that value, instead of just evaluating an expression, should be irrelevant to how we display the resulting value.

**Comment by valyagolev** [comment]
I still think it's good to mark it differently somehow, even if the particular syntax is questionable

**Comment by byorgey** [comment]
> I still think it's good to mark it differently somehow, even if the particular syntax is questionable

Why? What difference does it make?

**Comment by xsebek** [comment]
@valyagolev I understand that it could be useful to make it different in order to more easily debug this new feature.

A technically correct syntax would be:
```
it0 = 0
it1 <- return 1
```

Honestly the debug features I previously introduced were never useful after merging. But it may just be because that they required recompiling to enable - it would be better if we could e.g. turn them on in the Main Menu.

I am fine with either way, but consider that making the result look different could be more confusing (GHCi does not do it either) and will require more code. 🤷

**Comment by valyagolev** [comment]
No, it's not really about debugging this feature, rather about avoiding
loss of information in REPL. If I run `some_func`, I probably also want to
see what exactly it returned, value-wise and type-wise ([there's][1] a very
useful `:set +t` in ghci). Otherwise there's no visual distinction between
two fundamentally different results. I'm sure the output syntax can be
tweaked.

That said, if the general opinion is still to  hide the distinction between
`cmd a` and `a` in repl, I'll remove the extra code that shows it.

[1]:
https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/ghci.html#ghci-options

**Comment by xsebek** [comment]
@valyagolev but there is no distinction between the values _that you get_:
```
> return 1
it0: int = 1
> 1
it1: int = 1
> it0 == it1
it3: bool = true
```

Sure the code at _the REPL line_ has different type, but you can see that in the corner and from the perspective of later lines the results are the same. 🤔 

If you have some other difference in mind, please give an example where it would have an impact. I might be missing something. 😅

**Comment by byorgey** [comment]
To offer a bit more perspective: there are essentially two different things you can do at the REPL.  You can use it as a "comand center" where you enter instructions (i.e. things with a `cmd` type) to be executed.  Or, you can use it as a "calculator" where you enter expressions to be evaluated.  For convenience, both work by entering an expression at the prompt: we just process them differently depending on the type. Mostly, I think players will not even think about the distinction, though it could be useful to have some way to help them understand the difference.  However, both modes result in a value, and once you get a value as output there is no longer any difference in how those values may be used.  So I think displaying the result values with different syntax is not the right way to go about helping players understand the distinction between executing commands and evaluating expressions.  I believe it is more likely to make them think there is some difference in how they can or should use the values later, which is not the case, so it would be more confusing than helpful.

**Comment by byorgey** [comment]
Also, I'm very open to thoughts/suggestions on whether the distinction between execution and evaluation could be confusing to players, and how we might make it less so.

**Comment by valyagolev** [comment]
I pushed a commit to remove the discussed syntax thing

**Review by byorgey — APPROVED** [review]
This looks great, thanks @valyagolev for your hard work on this!

Now we just need to implement #60 so we can actually see the previous `it{n}` bindings!

**Comment by byorgey** [comment]
@valyagolev in the future remember you can merge by adding the `merge me` label, then the Mergify bot will do everything (including making sure that the branch still passes all the tests after merging the current state of the `main` branch).

**Comment by valyagolev** [comment]
oh right, sorry, forgot about that

**Pull request merged**
